partiql-logical-planner/src/typer.rs

```diff
@@ -0,0 +1,897 @@
+    derived_type: PartiqlType,
+}
+
+#[allow(dead_code)]
+impl TypeEnvContext {
+    fn new() -> Self {
+        TypeEnvContext::default()
+    }
+
+    fn env(&self) -> &LocalTypeEnv {
+        &self.env
+    }
+
+    fn derived_type(&self) -> &PartiqlType {
+        &self.derived_type
+    }
+}
+
+impl Default for TypeEnvContext {
+    fn default() -> Self {
+        TypeEnvContext {
+            env: LocalTypeEnv::new(),
+            derived_type: any!(),
+        }
+    }
+}
+
+impl From<(&LocalTypeEnv, &PartiqlType)> for TypeEnvContext {
+    fn from(value: (&LocalTypeEnv, &PartiqlType)) -> Self {
+        TypeEnvContext {
+            env: value.0.clone(),
+            derived_type: value.1.clone(),
+        }
+    }
+}
+
+/// Represents a Local Type Environment as opposed to the Global Type Environment in the Catalog.
+type LocalTypeEnv = IndexMap<SymbolPrimitive, PartiqlType>;
+
+#[derive(Debug, Clone)]
+pub struct PlanTyper<'c> {
+    typing_mode: TypingMode,
+    catalog: &'c dyn Catalog,
+    logical_plan: LogicalPlan<BindingsOp>,
+    errors: Vec<TypingError>,
+    type_env_stack: Vec<TypeEnvContext>,
+    bop_stack: Vec<BindingsOp>,
+    output: Option<PartiqlType>,
+}
+
+#[allow(dead_code)]
+impl<'c> PlanTyper<'c> {
+    /// Creates a new [PlanTyper] for the given Catalog and Intermediate Representation with `Strict` Typing Mode.
+    pub fn new_strict(catalog: &'c dyn Catalog, ir: &LogicalPlan<BindingsOp>) -> Self {
+        PlanTyper {
+            typing_mode: TypingMode::Strict,
+            catalog,
+            logical_plan: ir.clone(),
+            errors: Default::default(),
+            type_env_stack: Default::default(),
+            bop_stack: Default::default(),
+            output: None,
+        }
+    }
+
+    /// Creates a new [PlanTyper] for the given Catalog and Intermediate Representation with `Permissive` Typing Mode.
+    pub fn new_permissive(catalog: &'c dyn Catalog, lg: &LogicalPlan<BindingsOp>) -> Self {
+        PlanTyper {
+            typing_mode: TypingMode::Permissive,
+            catalog,
+            logical_plan: lg.clone(),
+            errors: Default::default(),
+            type_env_stack: Default::default(),
+            bop_stack: Default::default(),
+            output: None,
+        }
+    }
+
+    /// Returns the typing result for the Typer
+    pub fn type_plan(&mut self) -> Result<PartiqlType, TypeErr> {
+        let ops = self.sort()?;
+
+        for idx in ops {
+            let graph = self.to_stable_graph()?;
+            if let Some(binop) = graph.node_weight(idx) {
+                self.type_bindings_op(binop)
+            }
+        }
+
+        if self.errors.is_empty() {
+            Ok(self.output.clone().unwrap_or(undefined!()))
+        } else {
+            Err(TypeErr {
+                errors: self.errors.clone(),
+            })
+        }
+    }
+
+    fn type_bindings_op(&mut self, op: &BindingsOp) {
+        self.bop_stack.push(op.clone());
```

Comment:
Seems like cloning the entire `BindingsOp` could get expensive with large plans.

Reply:
`BindingOp` is not nested so I wouldn't  expect this to become expensive; good call-out though as made me to revisit the stack and see if we actually need it now. I removed it altogether.